ludwig/models/llm.py

```diff
@@ -0,0 +1,330 @@
+                a dictionary of input names to input tensors and targets is a
+                dictionary of target names to target tensors.
+            mask: A mask for the inputs.
+
+        Returns:
+            A dictionary of output {feature name}::{tensor_name} -> output tensor.
+        """
+
+        if isinstance(inputs, tuple):
+            inputs, targets = inputs
+            # Convert targets to tensors.
+            for target_feature_name, target_value in targets.items():
+                if not isinstance(target_value, torch.Tensor):
+                    targets[target_feature_name] = torch.from_numpy(target_value)
+                else:
+                    targets[target_feature_name] = target_value
+        else:
+            targets = None
+
+        assert list(inputs.keys()) == self.input_features.keys()
+
+        print("INPUTS", self.get_input_ids(inputs))
+        with torch.no_grad():
+            outputs = self.model.generate(
+                input_ids=self.get_input_ids(inputs),
+                attention_mask=mask,
+                generation_config=self.generation_config,
+                max_new_tokens=self.max_new_tokens,
+                return_dict_in_generate=True,
+                output_scores=True,
+            )
+        print("OUTPUTS", outputs)
+        return self.extract(inputs, outputs)
+
+    def get_input_ids(self, inputs):
+        """Returns the input ids for the text feature input."""
+        return inputs[self.config_obj.input_features[0].name].type(torch.int32)
+
+    def extract(
+        self,
+        inputs: Union[
+            Dict[str, torch.Tensor], Dict[str, np.ndarray], Tuple[Dict[str, torch.Tensor], Dict[str, torch.Tensor]]
+        ],
+        outputs,
+    ):
+        """Extracts predictions and probabilities from the model outputs."""
+        of_feature_type = self.config_obj.output_features[0].type
+        if of_feature_type == TEXT:
+            return {
+                self.config_obj.output_features[0].name: {
+                    "predictions": self.extract_text_predictions(inputs, outputs.sequences),
+                    # Unnormalized log probabilities
+                    # It is a tuple containing one entry for each generated token. Each tuple member is a tensor
+                    # containing the log probabilities from the model, for all words in the vocabulary.
+                    # "probabilities": self.extract_logits(outputs.scores),
+                }
+            }
+        if of_feature_type == CATEGORY:
+            predictions = self.extract_category_predictions(inputs, outputs.sequences)
+            return {
+                self.config_obj.output_features[0].name: {
+                    "predictions": predictions["predictions"],
+                    "probabilities": predictions["probabilities"],
+                }
+            }
+
+    def extract_text_predictions(self, inputs, output_sequences):
```

Comment:
This stuff should all live in the decoders.

Reply:
@tgaddair would you recommend creating a custom decoder for Text Output Features (LLMs specifically)?

Reply:
Yeah, I would.

Reply:
Makes sense, I agree it both makes more sense and will be exponentially cleaner. About halfway into it at the moment